src/mpid/common/sched/mpidu_sched.c

```diff
@@ -488,10 +494,12 @@ int MPIDU_Sched_start(MPIR_Sched_t * sp, MPIR_Comm * comm, int tag, MPIR_Request
 
     /* finally, enqueue in the list of all pending schedules so that the
      * progress engine can make progress on it */
-    if (all_schedules.head == NULL)
-        MPID_Progress_activate_hook(MPIR_Nbc_progress_hook_id);
 
+    MPID_THREAD_CS_ENTER(VCI, MPIDIU_THREAD_SCHED_LIST_MUTEX);
     DL_APPEND(all_schedules.head, s);
+    MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_SCHED_LIST_MUTEX);
+
+    MPID_Progress_activate_hook(MPIR_Nbc_progress_hook_id);
```

Comment:
Could we add a comment about why the check for `all_schedules.head == NULL` is removed? Would be useful for future readers.

Reply:
Actually, that explanation should go in the commit message.

Reply:
Also the order of calling `DL_Append` and `MPID_Progress_activate_hook` are changed. Were these changes  result of adding the critical section?

Reply:
Good that you spotted it :) Originally it has a check for `all_schedules.head==NULL` and only activate it when it is NULL. I removed that check since I reason that inside `MPID_Progress_activate_hook` will do a check anyway, so the check doesn't seem to save anything -- assuming the atomic load inside `MPID_Progress_activate_hook` is really cheap anyway. 

I swapped the order on the reasoning that between the activating hook and appending schedules, another thread may deactivate the hook in-between.

Does that make sense?

Reply:
I see. Thanks for explaining. Do you think we can treat these changes as orthogonal to the current patch (adding granular CS..) and move to a separate commit?